src/stc/stc.cpp.in

```diff
@@ -563,6 +564,41 @@ void wxStyledTextCtrl::AnnotationClearLine(int line) {
     SendMsg(SCI_ANNOTATIONSETTEXT, line, (sptr_t)NULL);
 }
 
+void wxBitmapToRGBA(const wxBitmap& bmp, unsigned char* rgba)
+{
+    const int totalPixels = bmp.GetWidth() * bmp.GetHeight();
+    wxImage img = bmp.ConvertToImage();
+    int curRGBALoc = 0, curDataLoc = 0, curAlphaLoc = 0;
+
+    for ( int i = 0; i < totalPixels; ++i )
+    {
+        rgba[curRGBALoc++] = img.GetData()[curDataLoc++];
+        rgba[curRGBALoc++] = img.GetData()[curDataLoc++];
+        rgba[curRGBALoc++] = img.GetData()[curDataLoc++];
+        rgba[curRGBALoc++] = img.HasAlpha()?img.GetAlpha()[curAlphaLoc++]:255;
+    }
+}
+
+void wxStyledTextCtrl::MarkerDefineBitmap(int markerNumber, const wxBitmap& bmp)
+{
+    // Store the current RGBAImage width and height.
+    const int oldRGBWidth = m_swx->GetRGBAImageWidth();
```

Comment:
I'm a bit confused by this: why do we need to save and restore this width/height? Is there a "current RGBA image" in STC?

Reply:
Scintilla keeps track of the width and height used for registering markers and images with RGBA data.  The specific variable used is the `Point sizeRGBAImage;` member from the editor class.

I thought it was important to save and restore for a scenario like this. Suppose someone has existing code registering 3 markers with 3 sets of 10x10 rgba data. 

```
m_stc->RGBAImageSetWidth(10);
m_stc->RGBAImageSetHeight(10);
m_stc->MarkerDefineRGBAImage(1,rgba1);
m_stc->MarkerDefineRGBAImage(2,rgba2);
m_stc->MarkerDefineRGBAImage(3,rgba3);
```

But suppose they later want to use a 12x12 png for the second marker. They don't want to go though the trouble of extracting the rgba data from the png, so they just load it into wxBitmap `bmp` and change the code to 

```
m_stc->RGBAImageSetWidth(10);
m_stc->RGBAImageSetHeight(10);
m_stc->MarkerDefineRGBAImage(1,rgba1);
m_stc->MarkerDefineBitmap(2,bmp);
m_stc->MarkerDefineRGBAImage(3,rgba3);
```

Since `MarkerDefineBitmap` internally calls `RGBAImageSetWidth/Height(12)`, the call `MarkerDefineRGBAImage(3,rgba3)` will now produce a garbled image (and possibly crash). Since a user might not expect `MarkerDefineBitmap` to change the current width and height used by `MarkerDefineRGBAImage`, I thought it best to save and restore them.

This is one of the reasons I was trying to come up with a way to pass a wxBitmap straight through in the `MarkerDefineRGBAImage`/`ResisterImage` methods. But never found a solutions that wasn't too ugly to use.

Reply:
I see, thanks for the explanation. It looks a bit weird to me to rely on width/height applying to more than just the next `DefineRGBAImage()` call, but it's better to be safe, thanks.